packages/slate/src/models/node.js

```diff
@@ -1346,20 +1332,57 @@ class Node {
    */
 
   getParent(key) {
+    if (key === this.key) return null
     if (this.hasChild(key)) return this
 
-    let node = null
+    const path = this.getPath(key)
+    if (path) {
+      return this.getDescendantAtPath(path.slice(0, -1))
+    }
+    return null
+  }
 
-    this.nodes.find(child => {
-      if (child.object == 'text') {
+  forEachDescendantWithPath(
+    iterator,
+    ancestorPath = [],
+    changedChildIndex = 0
+  ) {
+    let ret
+
+    this.nodes.forEach((child, i) => {
+      if (i < changedChildIndex) {
+        ret = true
+        return true
+      }
+
+      const path = [...ancestorPath, i]
+      if (iterator(child, path) === false) {
+        ret = false
         return false
-      } else {
-        node = child.getParent(key)
-        return node
+      }
+
+      if (child.object != 'text') {
+        ret = child.forEachDescendantWithPath(iterator, path)
+        return ret
       }
     })
 
-    return node
+    return ret
+  }
+
+  regeneratePathsCache() {
```

Comment:
I'm not keen on these types of methods being exposed, and generally what they do to the complexity of the code. Is there a way to restructure this to use the existing memoization technique, for top-down memoization?